### PR TITLE
feat: add more market user position alert types

### PR DESCRIPTION
**Pull request by OnlyJousting**
Changes:
- Adds "Fully converted" and "Incomplete conversion" alerts to user positions
- Adds new `getLiquidationStatus` method while marking old one as depreciated (still used in legacy forms)
- Syncs user position market page alerts with market-list user position tooltips

### CODE REVIEW

File: apps/main/src/llamalend/llamalend.types.ts [hunk 1]
```diff
@@ -11,6 +11,15 @@ export type NetworkDict<ChainId extends IChainId = IChainId> = Record<ChainId, L
 export type LlamaMarketTemplate = MintMarketTemplate | LendMarketTemplate
 
 export type HealthColorKey = 'healthy' | 'close_to_liquidation' | 'soft_liquidation' | 'hard_liquidation' | ''
+
+export type UserPositionStatus =
```
Comment (0xAlunara) [inline apps/main/src/llamalend/llamalend.types.ts:15]: Not that it matters I guess but could've deduced implicitly with `Required<ReturnType<typeof getLiquidationStatus>>`

File: apps/main/src/llamalend/position-status-content.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    severity: 'info',
+  },
+  softLiquidation: {
+    title: t`Liquidation protection active`,
+    description: (
+      <Trans>
+        The collateral price is inside your liquidation range. Within the range, collateral is being converted between{' '}
+        {collateralSymbol} and {borrowSymbol} as the price goes up or down. These conversions incur losses that reduce
+        the position's health.
+        <br />
```
Comment (0xAlunara) [inline apps/main/src/llamalend/position-status-content.tsx:36]: Reminder that you need a double <br /> for an actual whiteline / paragraph spacing. Personally I try to avoid br's and prefer gapped stacks with typography (or a span if you dont give a shit about typography)
I believe we've removed the default margins for <p> so that's not really an option I think

File: apps/main/src/llamalend/position-status-content.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+): Record<UserPositionStatus, PositionStatusContent> => ({
+  healthy: {
+    title: t`Position is healthy`,
+    description: (
+      <Trans>The oracle price is above your liquidation range. No soft-liquidation conversions are occurring.</Trans>
+    ),
+    severity: 'info',
+  },
+  closeToLiquidation: {
+    title: t`Position is healthy`,
```
Comment (0xAlunara) [inline apps/main/src/llamalend/position-status-content.tsx:23]: Double checking if this is really intended to equal `healthy` or a copy-paste error

File: apps/main/src/llamalend/position-status-content.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        <br />
+        <br />
+        If price recovers into your range, conversions resume and losses will further reduce your health. Consider
+        closing your position or repaying debt to improve health before that happens.
+      </Trans>
+    ),
+    severity: 'warning',
+  },
+  incompleteConversion: {
+    title: t`Position at risk — incomplete conversion`,
```
Comment (0xAlunara) [inline apps/main/src/llamalend/position-status-content.tsx:70]: No em-dash? 🤓

File: apps/main/src/llamalend/features/market-list/cells/MarketTitleCell/UserMarketPositionIndicator.tsx [hunk 5]
```diff
@@ -32,55 +24,29 @@ const flickerEffect = (
   return () => clearTimeout(timeoutId)
 }
 
-const Paragraph = ({ children }: { children?: ReactNode }) => (
-  <Typography variant="bodySRegular" fontWeight={FontWeight.Medium}>
```
Comment (0xAlunara) [inline apps/main/src/llamalend/features/market-list/cells/MarketTitleCell/UserMarketPositionIndicator.tsx:36]: Note that you're no longer using `bodySRegular` in the new tooltips, which means the (description) text will be a bit bigger.

**Review by 0xAlunara — DISMISSED** [review]

File: apps/main/src/llamalend/features/market-position-details/hooks/useBorrowPositionDetails.ts [hunk 6]
```diff
@@ -112,16 +112,23 @@ export const useBorrowPositionDetails = ({
...
-    return getLiquidationStatus(healthNotFullValue, isClose, borrowed ?? '0')
-  }, [hasLoan, healthNotFullValue, userBandsValue, liquidationBand, oraclePriceBand, borrowed])
+  const positionStatus = useMemo(() => {
+    if (!hasLoan || !healthNotFullValue || !userBandsValue) return undefined
+    const isCloseToSoftLiquidation = getIsUserCloseToSoftLiquidation(
+      userBandsValue[0],
+      liquidationBand ?? null,
+      oraclePriceBand,
+    )
+    // userBandsValue[1] is the lower price boundary of a users range. Band numbers go up in numbers when going down in prices
```
Comment (DanielSchiavini) [inline apps/main/src/llamalend/features/market-position-details/hooks/useBorrowPositionDetails.ts]: seems like an array destructuring like `const [upperBoundary, lowerBoundary] = userBandsValue`

File: apps/main/src/llamalend/features/market-position-details/BorrowPositionDetails.tsx [hunk 7]
```diff
@@ -80,8 +78,15 @@ export const BorrowPositionDetails = ({
   totalDebt,
 }: BorrowPositionDetailsProps) => (
   <Stack padding={Spacing.md} gap={Spacing.xs}>
-    {liquidationAlert.softLiquidation && <LiquidationAlert type="soft" />}
-    {liquidationAlert.hardLiquidation && <LiquidationAlert type="hard" />}
+    {liquidationAlert.status &&
+      liquidationAlert.status !== 'healthy' &&
```
Comment (DanielSchiavini) [inline apps/main/src/llamalend/features/market-position-details/BorrowPositionDetails.tsx]: this will break when we add more statuses, status that need alerts should be defined together with the status enum imo

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts [hunk 8]
```diff
@@ -70,11 +70,20 @@ export function useUserMarketStats(market: LlamaMarket, column?: LlamaMarketColu
 
   const borrowedAmount = stats ? ('borrowed' in stats ? stats.borrowed : stats.stablecoin) : 0
 
+  const status = stats
+    ? getLiquidationStatus(
+        decimal(stats.health) ?? '0',
```
Comment (DanielSchiavini) [inline apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: no position != 0 health. Users will be alerted for liquidation incorrectly

File: apps/main/src/llamalend/llama.utils.ts [hunk 9]
```diff
@@ -242,7 +260,11 @@ export function getLiquidationStatus(
   return userStatus
 }
 
-export function getIsUserCloseToLiquidation(
+// is user fully converted
```
Comment (DanielSchiavini) [inline apps/main/src/llamalend/llama.utils.ts]: what is this about?

File: apps/main/src/llamalend/features/market-position-details/BorrowPositionDetails.tsx (hunk 7 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/features/market-position-details/BorrowPositionDetails.tsx]: I was trying to solve it with types alone at first but adding a hasMarketAlert field to the object worked out better 🤝

File: apps/main/src/llamalend/llama.utils.ts (hunk 9 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/llama.utils.ts]: I forgot to clean them up, was considering making helpers

File: apps/main/src/llamalend/position-status-content.tsx (hunk 2 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/position-status-content.tsx:32]: Br seems a bit more declarative to me when there's a need for line breaks too

File: apps/main/src/llamalend/position-status-content.tsx (hunk 4 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/position-status-content.tsx]: ey I just copied the copy from the notion ticket, props for noticing though

File: apps/main/src/llamalend/position-status-content.tsx (hunk 3 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/position-status-content.tsx]: ah this is actually not a status we have designed for. it's used in the legacy use cases but not the new ones all though maybe it would be nice to have. I'll remove it for now though

File: apps/main/src/llamalend/llamalend.types.ts (hunk 1 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/llamalend.types.ts]: it's cleaner 👌

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts (hunk 8 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: omg they weren't even needed anymore, thanks for pointing it out

File: apps/main/src/llamalend/position-status-content.tsx (hunk 4 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/position-status-content.tsx]: actually, they are — not -?

File: apps/main/src/llamalend/features/market-list/cells/MarketTitleCell/UserMarketPositionIndicator.tsx (hunk 5 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/features/market-list/cells/MarketTitleCell/UserMarketPositionIndicator.tsx:36]: nice, I restored the typography variant

File: apps/main/src/llamalend/position-status-content.tsx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    description: (
+      <Trans>
+        The collateral price is inside your liquidation range. Within the range, collateral is being converted between{' '}
+        {collateralSymbol} and {borrowSymbol} as the price goes up or down. These conversions incur losses that reduce
+        the position's health.
+        <br />
+        You cannot add collateral while in the range. Close your position to recover current collateral. You can also
+        monitor health closely and repay debt to keep the position open.
+        <br />
+        If health reaches 0, your position is hard-liquidated.
```
Comment (0xPearce) [inline apps/main/src/llamalend/position-status-content.tsx]: Inconsistent with the description of hard liquidation 

```suggestion
        If health reaches 0, your position can be hard-liquidated.
```

File: apps/main/src/llamalend/llama.utils.ts [hunk 11]
```diff
@@ -212,6 +212,23 @@ const getSoftLiquidationThreshold = (userIsCloseToLiquidation: boolean) => (user
...
+  userIsCloseToSoftLiquidation: boolean,
+  userIsBelowRange: boolean,
+  userStateCollateral: Decimal | undefined,
+  userStateBorrowed: Decimal | undefined,
+) {
+  if (healthNotFull == null || userStateCollateral == null || userStateBorrowed == null) return undefined
+  const threshold = getSoftLiquidationThreshold(userIsCloseToSoftLiquidation)
+  if (+healthNotFull < 0) return 'hardLiquidation' as const
+  if (userIsBelowRange && +userStateCollateral > threshold) return 'incompleteConversion' as const
+  if (userIsBelowRange && +userStateCollateral <= threshold) return 'fullyConverted' as const
```
Comment (0xPearce) [inline apps/main/src/llamalend/llama.utils.ts]: we shouldn't use the same threshold for collateral !! we should use the usd value of the collateral? If the collateral is WBTC here, 0.05 WBTC would be considered as dust and the position as fully converted, even though it represent +3k $

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts [hunk 12]
```diff
@@ -70,11 +70,20 @@ export function useUserMarketStats(market: LlamaMarket, column?: LlamaMarketColu
 
   const borrowedAmount = stats ? ('borrowed' in stats ? stats.borrowed : stats.stablecoin) : 0
 
+  const status = stats
+    ? getLiquidationStatus(
+        decimal(stats.health),
+        stats.softLiquidation,
+        !stats.softLiquidation && borrowedAmount > 0,
```
Comment (0xPearce) [inline apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: for the `userIsBelowRange` how about the case that user can retain a small converted borrow token after they already exited soft liquidation? In that case positions that are back above range will be shown as fully converted

File: apps/main/src/llamalend/features/market-position-details/hooks/useBorrowPositionDetails.ts [hunk 13]
```diff
@@ -119,8 +119,13 @@ export const useBorrowPositionDetails = ({
...
     )
-    // userBandsValue[1] is the lower price boundary of a users range. Band numbers go up in numbers when going down in prices
-    const isUserBelowRange = oraclePriceBand && userBandsValue[1] ? oraclePriceBand > userBandsValue[0] : false
+    // Band numbers go up as prices go down (e.g. band 20 = 100-200, band 21 = 50-100), so the
+    // higher band number (userBandsValue[1]) is the lower price boundary of the user's range.
+    const [upperBoundary, lowerBoundary] = userBandsValue
+    const isUserBelowRange =
+      oraclePriceBand != null && upperBoundary != null && lowerBoundary != null
+        ? oraclePriceBand > lowerBoundary
+        : false
```
Comment (DanielSchiavini) [inline apps/main/src/llamalend/features/market-position-details/hooks/useBorrowPositionDetails.ts]:
```suggestion
      oraclePriceBand != null && lowerBoundary != null && oraclePriceBand > lowerBoundary
```

**Review by DanielSchiavini — DISMISSED** [review]

File: apps/main/src/llamalend/llama.utils.ts (hunk 11 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/llama.utils.ts]: ah ofc, I wasn't sure if threshold made sense to use anyways but this makes it clear that it's not

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts (hunk 12 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: the best would be to be able to look at current oracle band and user range band numbers like on the market page, maybe we can request this 🤔

File: apps/main/src/llamalend/position-status-content.tsx (hunk 4 above)
Comment (DanielSchiavini) [reply apps/main/src/llamalend/position-status-content.tsx]: We should be consistent, we use "-" everywhere

Comment (OnlyJousting) [reply apps/main/src/llamalend/position-status-content.tsx]: I'll change them to - then

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts (hunk 12 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: yea Benny is adding current oracle band to the response so we can simply check if we've passed the range

**Review by DanielSchiavini — DISMISSED** [review]

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts (hunk 12 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: added now

**Review by 0xtutti — DISMISSED** [review]

**Review by 0xAlunara — APPROVED** [review]

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts [hunk 14]
```diff
@@ -70,11 +70,20 @@ export function useUserMarketStats(market: LlamaMarket, column?: LlamaMarketColu
...
+      )
+    : undefined
+
   return {
     ...(stats && {
       data: {
-        softLiquidation: stats.softLiquidation,
-        liquidated: stats.healthFull < 0 && borrowedAmount === 0,
+        status,
         health: stats.healthFull,
```
Comment (Pearce542) [inline apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: we use `stats.health` to calculate the status, but then we display the `stats.healthFull` which can create a mismatch

File: apps/main/src/llamalend/features/market-list/cells/HealthCell.tsx [hunk 15]
```diff
@@ -15,13 +15,14 @@ const { Spacing } = SizesAndSpaces
...
+  const { health, status } = data ?? {}
+  const softLiquidation = status === 'softLiquidation'
   return health == null ? (
     error && <ErrorCell error={error} />
   ) : (
     <Tooltip
-      title={softLiquidation ? 'Liquidation Protection On' : 'Position active'}
-      body={<HealthTooltipContent softLiquidation={!!softLiquidation} />}
+      title={softLiquidation ? t`Liquidation Protection On` : t`Position active`}
+      body={<HealthTooltipContent softLiquidation={softLiquidation} />}
```
Comment (Pearce542) [inline apps/main/src/llamalend/features/market-list/cells/HealthCell.tsx]: We should handle the other statuses as well (fully converted, hard liquidation, incompleteConversion etc) otherwise the tooltip content is wrong. Or at least invert `const healthy = status === 'healthy'`

**Review by Pearce542 — COMMENTED** [review]
I don't know how you tested all the edge cases, or whether you're confident this will display the correct content in every case. We should absolutely add mocked tests like we did for the borrow and supply forms otherwise, we'll keep fixing one thing and breaking another.

File: apps/main/src/llamalend/features/market-list/cells/HealthCell.tsx (hunk 15 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/features/market-list/cells/HealthCell.tsx]: "Position active" is universal enough to not be incorrect but I guess we can just match the alerts and it would be even more accurate

**Comment by OnlyJousting** [comment]
> I don't know how you tested all the edge cases, or whether you're confident this will display the correct content in every case. We should absolutely add mocked tests like we did for the borrow and supply forms otherwise, we'll keep fixing one thing and breaking another.

I think now that we use current band to determine if the position is below together with a soft liquidation flag that it's pretty straight forward to determine what the correct state is but I can add mock tests too

File: apps/main/src/llamalend/queries/market-list/llama-market-stats.ts (hunk 14 above)
Comment (OnlyJousting) [reply apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: ah, stats.health is healthNotFull?

Comment (OnlyJousting) [reply apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: I'll double check all health things

Comment (OnlyJousting) [reply apps/main/src/llamalend/queries/market-list/llama-market-stats.ts]: nvm this is correct, you use healthNotFull to check if there's hard liquidation risk. the user table isn't utilizing both values to display health like we do in the market page though which might cause some confusion 🤔

**Review by 0xtutti — DISMISSED** [review]

**Review by 0xAlunara — DISMISSED** [review]

**Review by 0xAlunara — APPROVED** [review]

**Review by 0xtutti — APPROVED** [review]
